KEP-2258: Promote NodeLogQuery to beta

aravindhp opened the pull request:
- Promote NodeLogQuery to beta:
  - highlight that a periodic job was added
  - link to the NodeLogQuery kubelet plugin

<!-- link to the k/enhancements issue -->
- Issue link:
  - https://github.com/kubernetes/enhancements/issues/2258

BenTheElder commented on keps/sig-windows/2258-node-log-query/README.md:
```diff
@@ -231,29 +231,30 @@ On Windows node, the same kubelet service logs will queried by explicitly
-[e2e tests](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/windows).
+Tests have been added to [kubelet node](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/node/kubelet.go)
+e2e tests that are run as a daily periodic job:
+- https://testgrid.k8s.io/sig-windows-master-release#capz-master-windows-alpha-nodelogquery
```
this is windows? where's the linux job?

aravindhp replied:
This jobs runs tests against both Linux and Windows nodes.

aravindhp replied:
I have explicitly called this out in the KEP now. PTAL.

BenTheElder replied:
This job cannot be release blocking quality for GA, FYI.

Because Azure jobs are not on a community owned account (see release blocking jobs policy: https://github.com/kubernetes/sig-release/blob/master/release-blocking-jobs.md) ... and these jobs are actually at risk of being shut down long term as we shut down the google cluster that has the unknown third-party secret for the account. (Which we're not doing in community clusters, ongoing discussions in SIG Testing / K8s Infra)

aravindhp replied:
Noted @BenTheElder. BTW, this feature will never be enabled by default. So does it need to be part of the release blocking job? Or are you saying this has to be moved off the Azure jobs irrespective? This would affect other sig-windows e2e tests too.

cc: @marosset @jsturtevant

jsturtevant replied:
> and these jobs are actually at risk of being shut down long term as we shut down the google cluster that has the unknown third-party secret for the account as part of finishing migrating to CI fully on K8s Infra clusters. (In community clusters we're not letting third parties add arbitrary secrets unlike the old Google clusters, ongoing discussions in SIG Testing / K8s Infra)

Where can we find more information about this? This will affect the CAPZ and image-builder jobs as well. cc: @jackfrancis

BenTheElder replied:
> Where can we find more information about this? This will affect the CAPZ and image-builder jobs as well. cc: @jackfrancis

We've (SIG K8s Infra + Testing) been talking to various folks (eg lachlan, jeremy, bridgett) at multiple venues for O(years) now but concretely hammering something out is ongoing, previously we'd hoped to get an Azure account properly under the project instead of going that route, but for recent discussion see this doc from SIG Testing + K8s Infra meetings this year: https://docs.google.com/document/d/1erBhuCwY26d0UfPbzt8lEj6bYT2hOUKzc2j36YHVqfM/edit

Thanks for CC-ing more folks, but please let's discuss _that_ aspect in #sig-k8s-infra or similar, it's an off-topic FYI as far as this KEP is concerned.

Aside from that, more on-topic FYI: This job cannot be release blocking and has not been for a long time now so if so if you want to GA this feature later you're going to need a different job for release blocking coverage.

aravindhp replied:
Going to keep this thread on topic 😄 

>This job cannot be release blocking and has not been for a long time now so if so if you want to GA this feature later you're going to need a different job for release blocking coverage.

Understood. I will figure out next steps before we get to GA starting with the next release.

BenTheElder reviewed (commented):
Sections like "Rollout, Upgrade and Rollback Planning" and "Monitoring Requirements" are missing _all_ of the questions and answers and are required for PRR review for promotion to beta.

aravindhp reviewed (commented):
> Sections like "Rollout, Upgrade and Rollback Planning" and "Monitoring Requirements" are missing _all_ of the questions and answers and are required for PRR review for promotion to beta.

@BenTheElder I am a bit confused by this. When I read the template I see:
```
### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->
```
Isn't this when a feature goes from beta to release? Or am I missing something?

BenTheElder commented:
> Isn't this when a feature goes from beta to release? Or am I missing something?

No, when a feature is targeted to be released as beta. The levels are alpha/beta/GA, all of which can be released.
"release" isn't a tier, it's when a feature being promoted to alpha/beta/GA happens in a release.

BenTheElder commented on keps/sig-windows/2258-node-log-query/kep.yaml:
```diff
@@ -18,7 +18,7 @@ approvers:
 last-updated: 2023-05-02
 # The target maturity stage in the current dev cycle for this KEP.
-stage: alpha
+stage: beta
```
the milestones should also be updated below under `milestone`

aravindhp commented:
Thanks for the feedback @BenTheElder. I have addressed your comments. PS: the manual testing was done when I was working on writing e2e tests for the feature.

johnbelamaric commented on keps/prod-readiness/sig-windows/2258.yaml:
```diff
@@ -1,3 +1,5 @@
 alpha:
   approver: "@johnbelamaric"
+beta:
+  approver: "@deads2k"
```
Is there a specific reason to switch to David? We usually try to keep it the same just since the previous reviewer will have context already.

aravindhp replied:
There is no specific reason other than that David [pinged](https://kubernetes.slack.com/archives/CPNHUMN74/p1707166593892599) me on the k8s slack asking about the KEP. So I assumed PRR review transferred to him. Apologies for making that assumption. I have made you the approver.

johnbelamaric commented on keps/sig-windows/2258-node-log-query/README.md:
```diff
@@ -226,34 +226,32 @@ tests will not be possible for this feature.
+have not received any negative feedback from cluster administrators and
+developers who have enabled the feature.
+
+A [kubelet plugin](https://github.com/aravindhp/kubectl-node-logs) has been released
```
s/kubelet/kubectl/

aravindhp replied:
Fixed and also added a link to the Krew index.

aravindhp reviewed (commented):
Thanks for the review, @johnbelamaric. I have addressed your comments.

johnbelamaric commented on keps/sig-windows/2258-node-log-query/README.md:
```diff
@@ -287,15 +286,42 @@ a 404 will be returned.
-_This section must be completed when targeting beta graduation to a release._
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+A rollout or rollback will not fail either when enabling or disabling the feature gate.
+The feature has no impact on running workloads.
```
What we're asking about here is a rollout of the feature enablement. So, if your code panics, it will cause the kubelet to crash. Or, maybe it could cause kubelet to fail to come up, for example?

If they enable the feature gate and then kubelet starts failing, that would be a reason to rollback.

johnbelamaric commented on keps/sig-windows/2258-node-log-query/README.md:
```diff
@@ -287,15 +286,42 @@ a 404 will be returned.
+The feature has no impact on running workloads.
+
+###### What specific metrics should inform a rollback?
+There are no specific metrics that inform a rollback other than the feature gate option
```
A few possibilities: kubelet crashes? kubelet memory consumption goes really high (memory leak in the API, for example)? kubelet CPU utilization spikes.

johnbelamaric commented on keps/sig-windows/2258-node-log-query/README.md:
```diff
@@ -287,15 +286,42 @@ a 404 will be returned.
 
-_This section must be completed when targeting beta graduation to a release._
+###### How can an operator determine if the feature is in use by workloads?
+This feature does not affect any workloads
```
Ok, good point. The spirit of the question is to help a cluster admin answer the questions: "Will I mess someone up if I disable this? If so, can identify who and go talk to them first?".

So, is there a way to tell if anyone is using this? A metric or log entry?

johnbelamaric reviewed (commented):
Ok, a few updates needed but generally PRR looks good.

aravindhp commented:
> Ok, a few updates needed but generally PRR looks good.

Thanks for the feedback, @johnbelamaric. Your comments were insightful and I have addressed the, It also made me think more to update some of the other answers to the scalability questions. Please take a look.

BenTheElder reviewed (commented):
one question, this is looking good!

johnbelamaric commented:
looks good! Be sure to answer Ben's question, though.

/approve

aravindhp commented on keps/sig-windows/2258-node-log-query/README.md:
```diff
@@ -226,34 +226,33 @@ tests will not be possible for this feature.
+
+A [kubectl plugin](https://github.com/aravindhp/kubectl-node-logs) has been released
+and added to the Krew [index](https://github.com/kubernetes-sigs/krew-index/blob/master/plugins/node-logs.yaml)
+for querying the logs more elegantly instead of using raw API calls.
```
SIG-cli was involved in the design and the decision was to not make this a core command at any point. This was also discussed during a SIG-arch call to drive the feature to acceptance. PS: this will always be a feature that defaults to off.

soltysh replied:
> Have we discussed with SIG CLI if this might become a core command later?

Yes it was discussed, but given this functionality is off by default, and there are some security concerns wrt that, we've decided it's best to keep it optional and the same applies to kubectl plugin.

BenTheElder replied:
Ack, thanks!

aravindhp commented:
@jsturtevant @marosset @knabben please lgtm and remove the hold from a SIG-windows perspective

knabben commented:
Tests  added, +1 to promote to beta from SIG-Windows

/lgtm

knabben commented:
/cancel hold

knabben commented:
/hold cancel
